### PR TITLE
Feature/signal start event

### PR DESCRIPTION
**Pull request by nightBaker**
## Signal Start Event + Process Disable/Enable

This PR adds three related features:

### 1. Signal Start Events
- New `SignalStartEvent` domain activity that creates workflow instances when a signal is broadcast
- `SignalStartEventListenerGrain` manages per-signal registrations across process definitions
- Fan-out: a single signal can start instances across multiple deployed processes
- Mirrors the existing `MessageStartEvent` pattern

### 2. Persistence Migration (Breaking Change)
- **Breaking:** `MessageStartEventListenerState` persistence migrated from JSON column (`ProcessDefinitionKeys`) to a normalized `MessageStartEventRegistrations` join table
- New `SignalStartEventRegistrations` table for signal listener persistence
- Existing message start event registrations will be lost on upgrade — re-deploy affected workflows after upgrading
- See comment in `FleanModelConfiguration.cs` for details

### 3. Process Disable/Enable
- `ProcessDefinition.Disable()` / `Enable()` domain methods control whether a process accepts new instances
- Disabled processes: block `StartWorkflow`, unregister all start event listeners (timer/message/signal)
- Re-enabling re-registers all start event listeners
- Deploy preserves disabled state (redeploying a disabled process keeps it disabled)
- Web UI toggle button with visual indicator for disabled processes

### Test Coverage
- 6 signal start event integration tests (create, fan-out, redeploy, empty state, failure 400/500)
- 8 factory grain tests for disable/enable (block instances, preserve state, listener registration)
- 2 domain tests, 3 BPMN converter tests
- Manual test plans in `tests/manual/17-signal-start-event/` and `tests/manual/18-start-event-undeploy/`

### CODE REVIEW

**Comment by nightBaker** [comment]
## Review Feedback Addressed (commit ea135bd)

All changes verified: build clean (0 errors), all tests passing (509 passed across all projects).

### 1. Missing failure tests for SignalStartEvent (comment #4038781715)
**Fixed.** Added 4 failure tests in `SignalStartEventTests.cs` per CLAUDE.md requirement:
- `FailActivity_ShouldSetErrorState_WithGenericException` — verifies ErrorCode 500 and correct message
- `FailActivity_ShouldSetErrorState_WithBadRequestActivityException` — verifies ErrorCode 400 and correct message
- `FailActivity_ShouldTransitionToNextActivity` — verifies failed activity still transitions (end event also completed)
- `FailActivity_ShouldNotMergeVariables` — verifies no variables merged on failure

Tests use `TaskActivity` (requires external completion/failure) instead of `ScriptTask` to properly exercise the `FailActivity` code path in signal-start-event-triggered workflows.

### 2. SendSignal error path returns misleading 404 (comment #4038782767)
**Fixed.** Added `BroadcastFailed` and `StartEventFailed` boolean flags to `SendSignalResult`. When both operations throw exceptions, the result now carries the failure flags. The controller checks these flags and returns 500 with "Failed to deliver signal due to internal errors" instead of the misleading 404 "No subscription or start event found".

| Scenario | Before | After |
|----------|--------|-------|
| No listeners registered | 404 | 404 (unchanged) |
| Both operations threw exceptions | 404 (misleading) | **500** |
| One succeeded, one failed | 200 | 200 (unchanged) |

**Review by nightBaker — COMMENTED** [review]
## Code Review - PR #131 (Signal Start Event + Process Disable/Enable)

Comprehensive PR adding signal start events, persistence migration, and process disable/enable. Well-structured implementation that mirrors existing patterns. Here is my review:

### What I checked

- CLAUDE.md compliance across all changed files
- Domain, grain, persistence, API, and Web UI layers
- Test coverage adequacy
- BPMN converter registration
- Manual test plans

---

### Findings

**No blocking issues found.**

#### Suggestions

**1. Persistence: Consider using cascade delete instead of manual cleanup**
In both `EfCoreMessageStartEventListenerGrainStorage.ClearStateAsync` and `EfCoreSignalStartEventListenerGrainStorage.ClearStateAsync`, registrations are manually removed before deleting the parent. If the FK relationship is configured with cascade delete in `FleanModelConfiguration`, the manual removal would be unnecessary. Currently the model config does not configure the FK relationship explicitly, so cascade may not be in effect. Not blocking since the manual approach is correct and safe.

**2. Domain: `ProcessDefinition.Disable()`/`Enable()` are minimal**
Per CLAUDE.md: "Domain state methods must be self-contained. If a flag and a timestamp always change together, they belong in the same domain state method." The current `Disable()`/`Enable()` methods only flip `IsActive`. If you later add a `DisabledAt` timestamp or audit trail, these methods are the right place to put them. Good forward-looking design.

**3. Persistence migration is a breaking change - well documented**
The `FleanModelConfiguration.cs` comment clearly documents that existing message start event registrations are lost on upgrade. Good. The PR description also calls this out as a breaking change with a mitigation path (re-deploy affected workflows). 

**4. `StartEventRegistration.cs` naming diverged from plan**
The plan called for a shared `StartEventRegistration` record, but the implementation correctly uses separate `MessageStartEventRegistration` and `SignalStartEventRegistration` records with distinct FK column names (`MessageName` vs `SignalName`). This is better than the plan since it provides type safety and clearer table semantics. Good deviation.

**5. Test coverage is solid**
- 6 signal start event integration tests covering create, fan-out, redeployment unregister, empty state, and failure scenarios
- 8 factory grain tests for disable/enable including edge cases (block instances, preserve disabled on redeploy, unregister/re-register listeners)
- 2 domain tests for SignalStartEvent
- 3 BPMN converter tests including missing signalRef validation
- Failure tests verify both 400 and 500 error codes, transition after failure, and no variable merge on failure - per CLAUDE.md requirements

**6. CLAUDE.md compliance**
- All logging uses `[LoggerMessage]` source generators (EventIds 6002-6004, 7005-7009, 9200-9204)
- Domain methods (`Disable()`/`Enable()`) are self-contained
- `SignalStartEvent` domain activity follows `MessageStartEvent` pattern exactly
- BPMN converter registered in the correct location
- Manual test plans in `tests/manual/17-signal-start-event/` and `tests/manual/18-start-event-undeploy/`
- Web UI uses `WorkflowEngine`/`CommandService` directly (not API) per CLAUDE.md
- Fluent UI components use `IconStart` parameter correctly on `FluentButton`

#### Minor

**7.** `MessageStartEventListenerGrain.UnregisterProcess` now returns early with a log if `RemoveProcess` returns false (key not found). The log message says "Unregistered" which is slightly misleading - it could say "already unregistered" or "not found". Same pattern in `SignalStartEventListenerGrain`.

**8.** The `SendSignalResult` record includes `BroadcastFailed` and `StartEventFailed` boolean flags, which is a nice addition beyond the original plan - gives the API controller better error reporting (500 vs 404 distinction).

---

**Verdict: Looks good.** Clean implementation that follows existing patterns, good test coverage, well-documented breaking change. The persistence migration to join tables is a solid improvement over JSON serialization.

**Review by nightBaker — COMMENTED** [review]
## Follow-up Review: Commit 69c7066

This commit addresses several items from previous reviews — redundant writes, misleading log messages, persistence round-trip tests, and `[LoggerMessage]` migration in `Workflows.razor`. Good work overall, but one blocking issue.

### :red_circle: Blocking — must fix before merge

**1. Missing `Workflows.razor.cs` code-behind — build will fail**
`Workflows.razor` now calls `LogLoadWorkflowsError`, `LogProcessStarted`, `LogProcessStartFailed`, `LogProcessStartError`, and `LogToggleProcessError`, but the code-behind partial class file `Workflows.razor.cs` containing the `[LoggerMessage]` declarations does not exist on the remote branch. The `@inject ILogger<Workflows> Logger` was removed, so these methods must be defined somewhere. It appears the file was created locally but not committed. The project will not compile on the remote branch.

### :green_circle: What looks good

- **`AddProcess` now returns `bool`** — both `MessageStartEventListenerState` and `SignalStartEventListenerState` return `false` when the process is already registered, and the grains skip the `WriteStateAsync` call. Clean pattern, avoids redundant persistence writes.

- **Log message fix** — `LogProcessUnregistered` renamed to `LogProcessNotFound` with clarified message. More accurate for the case where the key wasn't in the list.

- **New `[LoggerMessage]` declarations** — EventIds 9105-9106 (message) and 9205-9206 (signal) are properly allocated and follow the existing pattern.

- **Persistence round-trip tests** — `SaveAndGetById_DisabledProcess_IsActiveFalseRoundTrip` and `SaveAndGetById_EnabledProcess_IsActiveTrueByDefault` verify the `IsActive` flag survives EF Core serialization. Good addition.

- **Test fix** — Duplicate activity instances removed from `SequenceFlow` constructors in `SignalStartEventTests`. Prevents potential object identity issues.

- **Design doc updated** — "Deploy preserves disabled state" now matches the implementation behavior.

**Comment by nightBaker** [comment]
## Addressed: Missing Workflows.razor.cs code-behind (Blocking)

Commit c3cd57e adds the missing `Workflows.razor.cs` code-behind file containing `[LoggerMessage]` source generator declarations (EventIds 1100-1104).

**What was wrong:** The file was created locally during the previous commit but was not staged/committed, so the remote branch would not compile.

**Verified:**
- Build passes (0 errors)
- All 148 tests pass

### Remaining non-blocking items from reviews (acknowledged):

1. **`SendSignal_ShouldFanOut_ToBothRunningAndStartEvents` uses `Task.Delay(500)`** - Noted, could be flaky on slow CI. Will address in a follow-up if it causes issues.
2. **PR description is empty** - Will add a summary below.

**Review by nightBaker — COMMENTED** [review]
## Code Review: Signal Start Event + Process Disable/Enable

Good work on this PR. The signal start event implementation correctly mirrors the existing message start event pattern, and the disable/enable feature is well-designed with proper defense-in-depth guards. The persistence migration, while breaking, is clearly documented.

### 🟢 Minor

1. **`ProcessDefinition.Disable()` / `Enable()` are trivially simple** — Currently they only toggle `IsActive`. Per CLAUDE.md "Domain state methods must be self-contained... if a flag and a timestamp always change together, they belong in the same domain state method." Right now there's no timestamp paired with the flag, so this is fine, but if `DisabledAt` / `EnabledAt` timestamps are added later, they should go inside these methods. Just a heads-up for future changes.

2. **`UnregisterAllStartEventListeners` / `RegisterAllStartEventListeners` use `FirstOrDefault` + null check** for looking up message/signal definitions. The `DeployWorkflow` refactored path uses the `IWorkflowDefinition` extension methods (`GetMessageStartEventNames()`, `GetSignalStartEventNames()`). Consider using those same extension methods inside the register/unregister helpers too, for consistency. Not blocking — the current code is correct.

3. **`SendSignalResult` gains `Errors` property** — Good addition for observability. Note that this changes the serialized shape of the API response. Existing clients that deserialize strictly may break on the new field. Low risk since it's nullable and defaults to null.

4. **`Workflows.razor.cs` code-behind** — Nice refactoring to use `[LoggerMessage]` source generators, properly compliant with CLAUDE.md logging conventions. The `@inject ILogger<Workflows> Logger` removal from the `.razor` file and replacement with `[Inject]` in the code-behind is clean.

### Overall Assessment

- **CLAUDE.md compliance**: ✅ — `[LoggerMessage]` used for all new log calls, domain methods encapsulate state transitions, DTOs use records, API endpoints in `WorkflowController`, Web UI uses grains directly via `WorkflowEngine`/`CommandService` (not HTTP API), Fluent UI `IconStart` pattern followed, manual test plan included with BPMN diagram section
- **Test coverage**: ✅ — 6 signal start event tests (completion + failure), 8 factory grain tests for disable/enable, persistence round-trip tests, domain tests
- **Architecture**: ✅ — Clean separation between domain (`Disable()`/`Enable()`), application (factory grain), and infrastructure (repository `UpdateAsync`). `IsProcessActive` guard in listener grains provides defense-in-depth
- **Known bug**: Issue #136 (DisableProcess/EnableProcess 500 error) was filed and fixed in a follow-up PR — good tracking

No blocking issues found. LGTM.

**Comment by nightBaker** [comment]
**Test 17 — Signal Start Event:** PASSED

- Deployed `signal-start-process` successfully (3 activities, 2 flows)
- `orderSignal` signal created instance `134164b7`, completed ✓
- Second `orderSignal` signal created separate instance `fd93d903`, completed ✓
- Unknown signal name returned 404 ✓
- All checklist items passed

*Note: Response includes `deliveredCount: 0` alongside `workflowInstanceIds` — this is expected behavior (deliveredCount refers to existing catch-event subscriptions, not new start-event instances).*

*Automated regression run by `manual-regression-testing` pipeline*

**Comment by nightBaker** [comment]
## Regression Fix: Timer cycle fires indefinitely (#137)

**Root cause:** `HandleBoundaryEventFired` always read the original `BoundaryTimerEvent.TimerDefinition` (e.g. `R3/PT5S`) from the immutable workflow definition. Each fire called `DecrementCycle()` on the same original, producing `R2/PT5S` every time — the count never reached zero.

**Fix:** Added `TimerCycleTracking` dictionary to `WorkflowInstanceState` that tracks the remaining cycle definition per `(hostActivityInstanceId, timerActivityId)`. A new `TimerCycleUpdated` domain event persists the decremented state after each fire. On subsequent fires, the aggregate reads from tracked state instead of the original definition.

**Changes:**
- `WorkflowInstanceState` — new `TimerCycleTracking` dictionary + `GetTimerCycleState`/`SetTimerCycleState` methods
- `WorkflowDomainEvents` — new `TimerCycleUpdated` event
- `WorkflowExecution.HandleBoundaryEventFired` — use tracked cycle state, emit `TimerCycleUpdated`
- `FleanModelConfiguration` — JSON column for `TimerCycleTracking` persistence
- 2 new tests: `ShouldDecrementAndStopAfterAllRepetitions` (R3 fires exactly 3 times) and `InfiniteCycle_ShouldAlwaysReRegister` (R/ fires indefinitely)

**Test results:** 301 domain tests + 157 application tests — all passing.

Fixes #137

**Comment by nightBaker** [comment]
**Test 04 — Parallel Gateway:** PASSED
Notes: branchA and branchB completed in separate scopes; join synchronized correctly.

**Comment by nightBaker** [comment]
**Test 06 — Call Activity:** PASSED
Notes: Parent completed; input=21, result=42; child instance created and completed.

**Comment by nightBaker** [comment]
**Test 09B — Message Boundary on IntermediateCatchEvent:** PASSED ✨
Notes: Previously a known bug — now FIXED. cancelBoundary fired correctly; cancelPath completed; cancelled=True.

**Comment by nightBaker** [comment]
**Test 10B — Signal Boundary on IntermediateCatchEvent:** PASSED ✨
Notes: Previously a known bug — now FIXED. emergencyStop signal fired SignalBoundaryEvent on longWait; longWait cancelled; emergencyPath completed; emergency=true.

**Comment by nightBaker** [comment]
**Test 11 — Error Boundary on CallActivity:** PASSED ✨
Notes: Previously a known bug — now FIXED. Parent completed (not failed); callFailing has ErrorState code=500; errorBoundary caught it; errorHandler ran; errorHandled=true; end2 reached. happyEnd not executed.

**Comment by nightBaker** [comment]
**Test 13B — Multi-Instance Parallel Cardinality:** PASSED
Notes: repeatTask ran 3×; results=['iter-0','iter-1','iter-2']; instance Completed.

**Comment by nightBaker** [comment]
## Final Regression Test Results — 2026-03-14

**Branch:** `feature/signal-start-event`
**Commit:** `9b7b66e` (Fix timer cycle firing indefinitely — #137)
**PR:** #131

### Full Regression Results

| # | Test | Result | Notes |
|---|------|--------|-------|
| 01 | Basic Workflow | **PASSED** | start, task1, end completed; done=True |
| 02 | Script Tasks | **PASSED** | 5 activities; x=15, greeting='hello' |
| 03 | Exclusive Gateway | **PASSED** | highTask taken; x=7; conditionalFlow=true |
| 04 | Parallel Gateway | **PASSED** | Both branches; join synchronized |
| 05 | Event-Based Gateway | **PASSED** | Message path won; timerCatch cancelled |
| 06 | Call Activity | **PASSED** | input=21, result=42; child completed |
| 07 | Embedded SubProcess | **PASSED** | subVar='from-subprocess'; parent continued |
| 08A | Timer Intermediate Catch | **PASSED** | timerFired=True after ~5s |
| 08B | Timer Boundary on ICE | **KNOWN BUG** | Pre-existing #129 |
| 09A | Message Catch | **PASSED** | Message delivered; afterApproval completed |
| 09B | Message Boundary on ICE | **PASSED** ✨ | Previously known bug — now FIXED |
| 10A | Signal Catch | **PASSED** | signalReceived=true; deliveredCount=1 |
| 10B | Signal Boundary on ICE | **PASSED** ✨ | Previously known bug — now FIXED |
| 11 | Error Boundary on CallActivity | **PASSED** ✨ | Previously known bug — now FIXED |
| 12 | Variable Scoping | **PASSED** | 3 isolated scopes confirmed |
| 13A | Multi-Instance Parallel Collection | **PASSED** | results=['processed-A','processed-B','processed-C'] |
| 13B | Multi-Instance Parallel Cardinality | **PASSED** | results=['iter-0','iter-1','iter-2'] |
| 13C | Multi-Instance Sequential | **PASSED** | results=['seq-X','seq-Y','seq-Z'] |
| 14A | Inclusive Gateway Parallel | **PASSED** | branch1+branch2 taken; branch3 skipped |
| 14B | Inclusive Gateway Default | **PASSED** | default flow taken; path='default' |
| 14C | Inclusive Gateway Nested | **PASSED** | inner/outer joins synced |
| 15A | Non-Interrupting Timer Boundary | **PASSED** | longTask active; reminderSent=true |
| 15B | Non-Interrupting Message Boundary | **PASSED** | longTask active; messageReceived=true |
| 15C | Timer Cycle R3/PT5S | **PASSED** ✨ | Bug #137 FIXED: fires exactly 3 times |
| 16 | Message Start Event | **PASSED** | 2 separate instances; unknown msg=404 |
| 17 | Signal Start Event | **PASSED** ✨ | Core feature: 2 instances; unknown signal=404 |
| 18 | Start Event Disable/Enable | **PASSED** ✨ | All 6 checklist steps pass |

### Summary

- **25 PASSED** (including 5 previously-known bugs now fixed: 09B, 10B, 11, 15C, 18)
- **1 KNOWN BUG** (pre-existing): 08B — timer boundary on IntermediateCatchEvent (#129)
- **0 new regressions**

### Positive Highlights

- **Bug #137 confirmed fixed** — Timer cycle `R3/PT5S` now fires exactly 3 times (not indefinitely)
- **09B, 10B fixed** — Message and signal boundary events on IntermediateCatchEvents now fire correctly
- **Test 11 fixed** — Error boundary on CallActivity catches child errors correctly
- **Test 17 PASSED** — Core new feature (Signal Start Event) works as expected
- **Test 18 PASSED** — Process disable/enable lifecycle fully functional (Bug #136 fix confirmed)

### Verdict

✅ **ALL TESTS PASSED** — moving to "Review by Human" for human review and merge.

---
*Automated regression run by `manual-regression-testing` pipeline — 2026-03-14*

File: src/Fleans/Fleans.Persistence/FleanModelConfiguration.cs [hunk 1]
```diff
@@ -38,6 +39,13 @@ public static void Configure(ModelBuilder modelBuilder)
 
             entity.Property(e => e.ProcessDefinitionId).HasMaxLength(512);
 
+            entity.Property(e => e.TimerCycleTracking)
+                .HasColumnName("TimerCycleTracking")
+                .HasConversion(
+                    v => JsonConvert.SerializeObject(v),
+                    v => JsonConvert.DeserializeObject<Dictionary<string, TimerDefinition>>(v)
```
Comment (nightBaker) [inline src/Fleans/Fleans.Persistence/FleanModelConfiguration.cs]: create separate class for TimerCycleTracking and store it to separate table

Comment (nightBaker) [reply src/Fleans/Fleans.Persistence/FleanModelConfiguration.cs]: Fixed in a20d6a4. Created `TimerCycleTrackingState` entity class with its own `TimerCycleTracking` table (composite key: `HostActivityInstanceId` + `TimerActivityId`). The JSON column on `WorkflowInstances` is replaced with a `HasMany` relationship with cascade delete. The `TimerDefinition` fields are stored as individual columns (`TimerType`, `TimerExpression`) instead of serialized JSON.
